chasm/tree.go

```diff
@@ -180,11 +187,52 @@ func (n *Node) CloseTransaction() (NodesMutation, error) {
 
 // Snapshot returns all nodes in the tree that have been modified after the given min versioned transition.
 // A nil minVT will be treated as the same as the zero versioned transition and returns all nodes in the tree.
-// This method should only be invoked when IsDirty() is false.
+// This method should only be invoked on root CHASM node when IsDirty() is false.
 func (n *Node) Snapshot(
-	minVT *persistencespb.VersionedTransition,
+	exclusiveMinVT *persistencespb.VersionedTransition,
 ) NodesSnapshot {
-	panic("not implemented")
+	if !softassert.That(n.logger, n.parent == nil, "chasm.Snapshot() should only be called on the root node") {
+		panic(fmt.Sprintf("chasm.Snapshot() called on child node: %+v", n))
+	}
+
+	nodes := make(map[string]*persistencespb.ChasmNode)
+	if err := n.snapshotInternal(exclusiveMinVT, []string{}, nodes); err != nil {
+		return NodesSnapshot{}
+	}
+
+	return NodesSnapshot{
+		Nodes: nodes,
+	}
+}
+
+func (n *Node) snapshotInternal(
+	exclusiveMinVT *persistencespb.VersionedTransition,
+	currentPath []string,
+	nodes map[string]*persistencespb.ChasmNode,
+) error {
+	if n == nil {
+		return nil
+	}
+
+	if transitionhistory.Compare(n.persistence.LastUpdateVersionedTransition, exclusiveMinVT) > 0 {
+		encodedPath, err := n.pathEncoder.Encode(n, currentPath)
+		if !softassert.That(n.logger, err == nil, "chasm path encoding should always succeed on clean tree") {
```

Comment:
I could return an error here and update all the callsite, but feels unnecessary as it should never happen. Tree is clean means we have already encoded & persisted all the paths.